lib/manager/components/FeedTransformationSettings.js

```diff
@@ -0,0 +1,440 @@
+                feedSource={feedSource}
+                onChange={this._onChangeTransformation}
+                onRemove={this._removeTransformation}
+                ruleSet={ruleSet}
+                transformation={t}
+                index={i} />
+            )
+          })
+          : <div className='lead'>No transformation steps defined.</div>
+        }
+        <ButtonToolbar>
+          <DropdownButton
+            title='Add step to transformation'
+            id='add-transformation-dropdown'
+            onSelect={this._addTransformation}>
+            {feedTransformationTypes.map(t => <MenuItem key={t} eventKey={t}>{getTransformationName(t)}</MenuItem>)}
+          </DropdownButton>
+        </ButtonToolbar>
+      </ListGroupItem>
+    )
+  }
+}
+
+class FeedTransformation extends Component<*, *> {
+  state = { csvData: null }
+
+  _getFieldsForType = (type: string) => {
+    const {feedSource, transformation} = this.props
+    // Get selected version (if applicable).
+    const version = feedSource.feedVersions &&
+      feedSource.feedVersions.find(v => v.id === transformation.sourceVersionId)
+    const fields = []
+    let index = 0
+    switch (type) {
+      case 'ReplaceFileFromVersionTransformation':
+        fields.push(
+          <VersionSelectorDropdown
+            key={index++}
+            dropdownProps={{
+              id: 'merge-versions-dropdown',
+              onSelect: this._onSelectVersion
+            }}
+            title={version
+              ? `From source version ${version.version}`
+              : 'Select a source version'
+            }
+            version={version}
+            versions={feedSource.feedVersions}
+          />
+        )
+        break
+      case 'ReplaceFileFromStringTransformation':
+        const inputIsUnchanged = this.state.csvData === null
+        const csvData = inputIsUnchanged
+          ? transformation.csvData
+          : this.state.csvData
+        const textValue = csvData || ''
+        const numLines = !textValue ? 0 : textValue.split(/\r*\n/).length
+        fields.push(
+          <div>
+            <label>Add the CSV data to add to/replace in the incoming GTFS:</label>
+            <textarea
+              onChange={this._onChangeCsvData}
+              style={{
+                fontFamily: 'monospace',
+                fontSize: 'x-small',
+                height: '80px',
+                overflow: 'auto',
+                width: '400px',
+                whiteSpace: 'pre'
+              }}
+              placeholder={
+                `stop_id,stop_code,stop_name,stop_lat,stop_lon
+1234567,188390987,Broad Ave,33.98768,-87.72686`
+              }
+              value={textValue} />
+            <div style={{marginBottom: '10px'}}>
+              <Button
+                bsSize='xsmall'
+                disabled={inputIsUnchanged}
+                style={{marginRight: '5px'}}
+                onClick={this._onSaveCsvData}>
+                Save CSV
```

Comment:
Why do we need to have a Save button for the text of this transform? Is the intent for the user to be able to discard changes if they made an edit by mistake?

Reply:
Because it's a free-form text input. Making a server request to save the input every time a keypress happens is not desirable. See deployment router and build config forms.